In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.svm import LinearSVC
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn import svm
from sklearn.metrics import precision_score, recall_score, accuracy_score, confusion_matrix, roc_auc_score
import pickle

In [2]:
X_train = pd.read_pickle('X_train_new.pickle')
X_test = pd.read_pickle('X_test_new.pickle')

In [3]:
X_train.dropna(inplace= True)
X_test.dropna(inplace= True)

In [4]:
X_train['tokenized_text'] = X_train['tokenized_text'].apply(lambda x: x.split())
X_test['tokenized_text'] = X_test['tokenized_text'].apply(lambda x: x.split())

In [5]:
X_train = X_train.sample(frac=0.02)

In [6]:
def identity_tokenizer(text):
    return text

tfidf_no_token = TfidfVectorizer(tokenizer=identity_tokenizer, preprocessor=identity_tokenizer,token_pattern=None, max_features = 400)

tfidf = TfidfVectorizer(
    analyzer='word',
    tokenizer=identity_tokenizer,
    preprocessor=identity_tokenizer,
    token_pattern=None) 

In [7]:
x_train = X_train['tokenized_text']
x_test = X_test['tokenized_text']
y_train = X_train['mental_state']
y_test = X_test['mental_state']

In [8]:
tfidf_no_token.fit(x_train)
fit_model = tfidf_no_token.transform(x_train)
test_model = tfidf_no_token.transform(x_test)

In [9]:
%%time
clf = SVC(probability=True, kernel='rbf', verbose = True)
clf.fit(fit_model, y_train)

[LibSVM]CPU times: user 8min 32s, sys: 3.94 ms, total: 8min 32s
Wall time: 8min 32s


SVC(probability=True, verbose=True)

In [10]:
def performance(y_true, y_hat):
    
    y_hat = list(map(lambda x: np.argmax(x), y_hat))
    
    print('-'*40)
    # accuracy
    print('Accuracy: ', accuracy_score(y_true,y_hat))
    # confusion matrix
    print('\n')
    print('Confusion Matrix: \n', confusion_matrix(y_true,y_hat))
    print('\n')
    # precision score of the model 
    #print('Precision: ', precision_score(y_true, y_hat, average='micro'))
    # recall score of the model 
    #print('Recall: ', recall_score(y_true, y_hat, average='micro'))
    # area under the ROC curve
    #print('Area under ROC curve: ', roc_auc_score(y_true, y_hat))
    print('-'*40)

In [11]:
predictions = clf.predict_proba(test_model)

In [12]:
performance(y_test, predictions)

----------------------------------------
Accuracy:  0.718972245722239


Confusion Matrix: 
 [[ 89591  15447   2924]
 [ 16262 131499  20364]
 [  9588  38392  42363]]


----------------------------------------


----------------------------------------
Accuracy:  0.7098081488961057


Confusion Matrix: 
 [[ 88509  16388   3065]
 [ 16889 130434  20802]
 [ 10313  38878  41152]]


----------------------------------------
sample train size = 1%

----------------------------------------
Accuracy:  0.718972245722239


Confusion Matrix: 
 [[ 89591  15447   2924]
 [ 16262 131499  20364]
 [  9588  38392  42363]]


---------------------------------------
sample train size = 2%